connector/connect/common/src/main/protobuf/spark/connect/relations.proto

```diff
@@ -781,3 +782,10 @@ message FrameMap {
   CommonInlineUserDefinedFunction func = 2;
 }
 
+message WithSequenceColumn {
```

Comment:
How is this different from Project(*, sequence_col), or WithColumn(sequence_col)?

Reply:
`withSequenceColumn` allows us to simply add a new column containing numbers increment from 1 regardless of the existing column whereas `withColumn` is not.

Let me introduce a brief example to help understanding.

Let's say we have a DataFrame as below:
```python
data = [("Alice", 35), ("Bob", 20), ("Charlie", 25)]
df = spark.createDataFrame(data, ["Name", "Age"])
df.show()
+-------+---+
|   Name|Age|
+-------+---+
|  Alice| 35|
|    Bob| 20|
|Charlie| 25|
+-------+---+
```

And now we want to add a column named "id", which has incremental number from 1.

If we use `withColumn`, we might need to use combination of `row_number` and `Window` as below:
```python
from pyspark.sql import Window
from pyspark.sql.functions import row_number

windowSpec = Window.orderBy("Age")
new_df = df.withColumn("id", row_number().over(windowSpec))

new_df.show()
+-------+---+---+
|   Name|Age| id|
+-------+---+---+
|    Bob| 20|  1|
|Charlie| 25|  2|
|  Alice| 35|  3|
+-------+---+---+
```

A "id" that increments from 1 was added, but the order of the existing data was changed.

We should kept the existing order to match the behavior with pandas.

There is another way to add an incrementing column without changing the order of the existing data is to use `monotonically_increasing_id` as below:
```python
from pyspark.sql.functions import monotonically_increasing_id

new_df = df.withColumn("id", monotonically_increasing_id() + 1)
+-------+---+---+
|   Name|Age| id|
+-------+---+---+
|  Alice| 35|  1|
|    Bob| 20|  2|
|Charlie| 25|  3|
+-------+---+---+
```

Well, the current result looks good, but unfortunately we can't guarantee that added columns always started from 1, because in fact the `monotonically_increasing_id` doesn't guarantee a starting value.

`withSequenceColumn` allows us to do this in very simple way as below:
```
new_df = df._withSequenceColumn("id")
new_df.show()
+---+-------+---+
| id|   Name|Age|
+---+-------+---+
|  0|  Alice| 35|
|  1|    Bob| 20|
|  2|Charlie| 25|
+---+-------+---+
```

Reply:
> How is this different from Project(*, sequence_col), or WithColumn(sequence_col)?

it was implemented as a logical plan, we don't have such an expression  cc @HyukjinKwon

Reply:
Yeah, the reason is that it requires to run a plan-wise operation `RDD.zipWithIndex`

Reply:
Well my argument against this is that it is just a project with a specific type of expression attached to it. There is no need to complicate the protocol, it is just that.

As for what this does, and this comment is more aimed at the original PR, two things:
- The IDs are not stable at all. This is an order based, and well we basically do not guarantee that the order is stable during processing.
- The IDs can contain gaps or duplicates if any of the shuffles of the input contains a non-deterministic column, and a retry of one of the input tasks/stages occurs. This is a result of the double scanning that RDD.zipWithIndex requires.
- Finally two scans can be slow.

Reply:
To make it as an expression, it would require to switch to a physical plan under the hood anyway, and this plan is more just internal - users are not supposed to directly call it.

The order doesn't matter. It's just to have a sequence to match with pandas'. Pandas API on Spark doesn't guarantee the order.

> This is a result of the double scanning that RDD.zipWithIndex requires.

If that's the case, I believe we should fix `RDD.zipWithIndex` to work properly.

Reply:
This PR adds just a protobuf message for the SQL plan that's internally used. I believe making it better is orthogonal in any event.

Reply:
Thank you for your feedback! Let me keep this PR as draft and will revisit after perform the refactoring of the existing functions first.

Reply:
https://github.com/apache/spark/pull/40456